extensions/data-transfer/portability-data-transfer-backblaze/src/main/java/org/datatransferproject/datatransfer/backblaze/common/BackblazeDataTransferClient.java

```diff
@@ -54,196 +55,271 @@
-          "User's credentials or permissions are not valid for any regions available", s3Exception);
+        bucketName = getOrCreateBucket(s3Client, listBucketsResponse, userRegion, exportService);
     }
 
-    bucketName = getOrCreateBucket(s3Client, listBucketsResponse, userRegion, exportService);
-  }
+    public String uploadFile(String fileKey, File file) throws IOException {
+        if (s3Client == null || bucketName == null) {
+            throw new IllegalStateException("BackblazeDataTransferClient has not been initialised");
+        }
+
+        try {
+            long contentLength = file.length();
+            monitor.debug(
+                    () -> String.format("Uploading '%s' with file size %d bytes", fileKey, contentLength));
+
+            if (contentLength >= sizeThresholdForMultipartUpload) {
+                monitor.debug(
+                        () ->
+                                String.format(
+                                        "File size is larger than %d bytes, so using multipart upload",
+                                        sizeThresholdForMultipartUpload));
+                return uploadFileUsingMultipartUpload(fileKey, file, contentLength);
+            }
+
+            PutObjectRequest putObjectRequest =
+                    PutObjectRequest.builder().bucket(bucketName).key(fileKey).build();
+
+            PutObjectResponse putObjectResponse =
+                    s3Client.putObject(putObjectRequest, RequestBody.fromFile(file));
 
-  public String uploadFile(String fileKey, File file) throws IOException {
-    if (s3Client == null || bucketName == null) {
-      throw new IllegalStateException("BackblazeDataTransferClient has not been initialised");
+            return putObjectResponse.versionId();
+        } catch (AwsServiceException | SdkClientException e) {
+            throw new IOException(String.format("Error while uploading file, fileKey: %s", fileKey), e);
+        }
     }
 
-    try {
-      long contentLength = file.length();
-      monitor.debug(
-          () -> String.format("Uploading '%s' with file size %d bytes", fileKey, contentLength));
+    /**
+     * Retrieves the account region from Backblaze B2 API using the provided credentials.
+     * This method implements exponential backoff retry logic for handling transient server errors.
+     *
+     * @param httpClient The HTTP client to use for making the API request
+     * @param keyId The Backblaze B2 account key ID
+     * @param applicationKey The Backblaze B2 application key
+     * @return The region string extracted from the s3ApiUrl in the response
+     * @throws BackblazeCredentialsException if:
+     *     - Authentication fails (4xx errors)
+     *     - Server errors persist after all retry attempts
+     *     - Response parsing fails
+     *     - Network errors occur and persist after retries
+     */
+    private String getAccountRegion(HttpClient httpClient, String keyId, String applicationKey)
```

Comment:
Looks like my formatter changed a lot of this file, but this function is the only new addition